dimod/binary/pybqm.py

```diff
@@ -0,0 +1,336 @@
+        if self.vartype == vartype:
+            return self
+
+        if vartype == Vartype.BINARY:
+            lin_mp = 2.
+            lin_offset_mp = -1.
+            quad_mp = 4.
+            lin_quad_mp = -2.
+            quad_offset_mp = .5
+        elif vartype == Vartype.SPIN:
+            lin_mp = .5
+            lin_offset_mp = .5
+            quad_mp = .25
+            lin_quad_mp = .25
+            quad_offset_mp = .125
+        else:
+            raise RuntimeError("unexpected vartype")
+
+        adj = self._adj
+
+        for u, Nu in adj.items():
+            lbias = Nu[u]
+
+            self.offset += lin_offset_mp * lbias
+            Nu[u] = lin_mp * lbias
+
+            for v, qbias in Nu.items():
+                if v == u:
+                    continue
+                Nu[v] = quad_mp * qbias
+                Nu[u] += lin_quad_mp * qbias  # linear
+                self.offset += quad_offset_mp * qbias
+
+        self._vartype = vartype
+
+        return self
+
+    def degree(self, v: Variable) -> int:
+        try:
+            return len(self._adj[v]) - 1
+        except KeyError:
+            raise ValueError(f"unknown variable {v!r}") from None
+
+    def energies(self, samples_like, dtype: DTypeLike = None):
+        samples, labels = as_samples(samples_like)
+
+        if len(labels) != self.num_variables():
+            raise ValueError("variable_order does not match the number of "
+                             "variables")
+
+        ldata = np.asarray([self.get_linear(v) for v in labels])
+
+        label_to_idx = {v: idx for idx, v in enumerate(labels)}
+        irow = []
+        icol = []
+        qdata = []
+        for u, v, bias in self.iter_quadratic():
+            irow.append(label_to_idx[u])
+            icol.append(label_to_idx[v])
+            qdata.append(bias)
+
+        irow = np.asarray(irow)  # type: ignore[assignment]
+        icol = np.asarray(icol)  # type: ignore[assignment]
+        qdata = np.asarray(qdata)  # type: ignore[assignment]
+
+        energies = samples.dot(ldata)
+        energies += (samples[:, irow]*samples[:, icol]).dot(qdata)
+        energies += ldata.dtype.type(self.offset)
+
+        return np.asarray(energies, dtype=dtype)
+
+    def get_linear(self, v: Variable) -> Any:
+        try:
+            return self._adj[v][v]
+        except KeyError:
+            raise ValueError(f"unknown variable {v!r}") from None
+
+    def get_quadratic(self, u: Variable, v: Variable,
+                      default: Optional[Any] = None) -> Any:
+        if u == v:
+            raise ValueError(f"{u!r} cannot have an interaction with itself")
+
+        try:
+            return self._adj[u][v]
+        except KeyError:
+            if default is None:
+                raise ValueError(
+                    f"{u!r} and {v!r} have no interaction") from None
+            return default
+
+    def is_linear(self) -> bool:
+        return not any(len(Nv) > 1 for Nv in self._adj.values())
```

Comment:
Looks like a :bug:.

Consider `dimod.BQM.from_ising({}, {'ab': 1, 'cd': 1})`. Has quadratic biases, but this function would return true.

Reply:
In adjdict we *always* keep the linear in the same dict, that's why we use `> 1`. I did add an extra test just to be sure :smile: